lib/astarte_device_sdk/mapping.c

```diff
@@ -93,3 +159,34 @@ astarte_result_t astarte_mapping_check_individual(
 
     return ASTARTE_RESULT_OK;
 }
+
+/************************************************
+ *         Static functions definitions         *
+ ***********************************************/
+
+static astarte_result_t check_path_segment(const char *endpoint_start, const char *endpoint_end,
+    const char *path_start, const char *path_end)
+{
+    size_t endpoint_segment_len = endpoint_end - endpoint_start;
+    size_t path_segment_len = path_end - path_start;
+    // The check will differ is the segment is a parameter or not
+    if ((strncmp(endpoint_start, "%{", 2) == 0)
+        && (endpoint_start[endpoint_segment_len - 1] == '}')) {
+        if (path_segment_len == 0) {
+            return ASTARTE_RESULT_MAPPING_PATH_MISMATCH;
+        }
+        for (size_t i = 0; i < path_segment_len; i++) {
+            if (path_start[i] == '#' || path_start[i] == '+') {
+                return ASTARTE_RESULT_MAPPING_PATH_MISMATCH;
+            }
+        }
+    } else {
+        if (endpoint_segment_len != path_segment_len) {
+            return ASTARTE_RESULT_MAPPING_PATH_MISMATCH;
+        }
+        if (memcmp(endpoint_start, path_start, path_segment_len) != 0) {
+            return ASTARTE_RESULT_MAPPING_PATH_MISMATCH;
+        }
```

Comment:
What do you think to use a single `if`?
```suggestion
        if ((endpoint_segment_len != path_segment_len) || (memcmp(endpoint_start, path_start, path_segment_len) != 0) {
            return ASTARTE_RESULT_MAPPING_PATH_MISMATCH;
        }
```